zcash_client_backend/src/tor/http/exchange_rate.rs

```diff
@@ -0,0 +1,150 @@
+use futures_util::{join, FutureExt};
+use rand::{seq::IteratorRandom, thread_rng};
+use serde::{Deserialize, Deserializer};
+use tracing::{error, trace};
+
+use crate::tor::{Client, Error};
+
+impl Client {
+    /// Fetches the latest USD/ZEC exchange rate.
+    ///
+    /// We query the ticker price of the following trading pairs:
+    /// - `ZECUSDT` on Binance
+    /// - `ZEC-USD` on Coinbase
+    /// - `zecusd` on Gemini
+    ///
+    /// Returns an error if all three exchange queries fail.
+    pub async fn get_exchange_rate_usd(&self) -> Result<f64, Error> {
+        // "Never go to sea with two chronometers; take one or three."
+
+        let binance_ticker = self
+            .get_json::<BinanceTicker>(
+                "https://api.binance.com/api/v3/ticker/price?symbol=ZECUSDT"
+                    .parse()
+                    .unwrap(),
+            )
+            .map(|res| res.map(|resp| resp.into_body()));
+
+        let coinbase_ticker = self
+            .get_json::<CoinbaseTicker>(
+                "https://api.exchange.coinbase.com/products/ZEC-USD/ticker"
+                    .parse()
+                    .unwrap(),
+            )
+            .map(|res| res.map(|resp| resp.into_body()));
```

Comment:
I have yet to see this ever succeed; so far in testing it always returns 403 Forbidden. Coinbase must be fetching a list of Tor exit nodes to block. I haven't checked whether providing an API key enables bypassing this.

If we can find an alternate ticket to use, we may want to consider dropping the Coinbase source.

Reply:
I've replaced Coinbase with MEXC; they appear to have similar volumes over the last 24 hours.

Reply:
I've now re-added Coinbase to the mix; we'll query them along with every other data source, just in case they start working.